extensions/amp-autocomplete/0.1/autocomplete-binding-single.js

```diff
@@ -0,0 +1,148 @@
+import {userAssert} from '../../../src/log';
+
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Single implementation of autocomplete. This supports autocompleting
+ * a single input value in its entirety.
+ * @implements {./autocomplete-binding-def.AutocompleteBindingDef}
+ * @private
+ */
+export class AutocompleteBindingSingle {
+  /**
+   * @param {!AMP.BaseElement} element
+   */
+  constructor({element}) {
+    /** @private {!Element} */
+    this.element_ = element;
+
+    /**
+     * The Single implementation of autocomplete will highlight
+     * the diff between the user input and the active suggestion.
+     * See displayActiveItemInInput() for more.
+     * @private {boolean}
+     */
+    this.shouldHighlight_ = element.hasAttribute('suggest-first');
+
+    /**
+     * The Single implementation of autocomplete will allow form
+     * submission with selection when "submit-on-enter" is present.
+     * @private {boolean}
+     */
+    this.submitOnEnter_ = element.hasAttribute('submit-on-enter');
+  }
+
+  /**
+   * Always try to autocomplete.
+   * @param {!HTMLInputElement} unusedInputEl
+   * @return {boolean}
+   */
+  shouldAutocomplete(unusedInputEl) {
+    return true;
+  }
+
+  /**
+   * The user input for generating suggestions is the full entry in the input element so far.
+   * @param {?HTMLInputElement} inputEl
+   * @return {string}
+   */
+  getUserInputForUpdate(inputEl) {
+    return inputEl.value || '';
+  }
+
+  /**
+   * Returns the full selection.
+   * @param {string} selection
+   * @param {!HTMLInputElement} unusedInputEl
+   * @param {string} unusedInput
+   * @return {string}
+   */
+  getUserInputForUpdateWithSelection(selection, unusedInputEl, unusedInput) {
+    return selection;
+  }
+
+  /**
+   * Input element value should be reset to the partially recorded user input.
+   * @param {string} userInput
+   * @param {!HTMLInputElement} inputEl
+   */
+  resetInputOnWrapAround(userInput, inputEl) {
+    inputEl.value = userInput;
+  }
+
+  /**
+   * Should only abide by "suggest-first" attribute if "filter" is "prefix".
+   * @return {boolean}
+   */
+  shouldSuggestFirst() {
+    const hasSuggestFirst = this.element_.hasAttribute('suggest-first');
+    const filter = this.element_.getAttribute('filter');
+    userAssert(
```

Comment:
Would multiple calls to `shouldSuggestFirst()` cause the component to yell multiple times? Can we check during binding construction instead?

Reply:
Good idea, moved. :)